File: 2015/src/main/java/nl/sbmf21/aoc15/days/Day2.java
package nl.sbmf21.aoc15.days;

import nl.sbmf21.aoc.common.ADay;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class Day2 extends ADay {

    private final ArrayList<Dimension> dimensions = parseDimensions();

    @Override
    public @NotNull Integer part1() {
        var total = 0;

        for (var dimension : dimensions) {
            total += dimension.size() + smallest(dimension.sides());
        }

        return total;
    }

    @Override
    public @NotNull Integer part2() {
        var total = 0;

        for (var dimension : dimensions) {
            total += dimension.volume() + smallest(dimension.perimeters());
        }

        return total;
    }

    public ArrayList<Dimension> parseDimensions() {
        var dimensions = new ArrayList<Dimension>();

        getInput().forEach(line -> dimensions.add(Dimension.fromLine(line)));

        return dimensions;
    }

    public int smallest(int[] values) {
        var smallest = values[0];

        for (int i = 1; i < values.length; i++) {
            var value = values[i];

            if (value < smallest) {
                smallest = value;
            }
        }

        return smallest;
    }

    public static class Dimension {

        public final int l, w, h;

        public static Dimension fromLine(String line) {
            var parts = line.split("x");

            return new Dimension(
                Integer.parseInt(parts[0]),
                Integer.parseInt(parts[1]),
                Integer.parseInt(parts[2])
            );
        }

        private Dimension(int l, int w, int h) {
            this.l = l;
            this.w = w;
            this.h = h;
        }

        public int[] sides() {
            return new int[]{
                l * w,
                w * h,
                h * l
            };
        }

        public int[] perimeters() {
            return new int[]{
                2 * (l + w),
                2 * (w + h),
                2 * (h + l)
            };
        }

        public int volume() {
            return l * w * h;
        }

        public int size() {
            var size = 0;

            for (var side : sides()) {
                size += side * 2;
            }

            return size;
        }
    }
}

File: 2021/kotlin/src/main/kotlin/nl/sbmf21/aoc21/days/Day2.kt
package nl.sbmf21.aoc21.days

import nl.sbmf21.aoc.common.ADay

class Day2(input: List<String>) : ADay(input) {

    private val actions = input.map { it.split(' ') }.map { Pair(it[0], it[1].toInt()) }

    override fun part1(): Int {
        var x = 0; var y = 0
        run({ x += it }, { y -= it }, { y += it }).also { return x * y }
    }

    override fun part2(): Int {
        var x = 0; var y = 0; var p = 0
        run({ x += it; y += (p * it) }, { p -= it }, { p += it }).also { return x * y }
    }

    private fun run(forward: (Int) -> Unit, up: (Int) -> Unit, down: (Int) -> Unit) = actions.forEach {
        when (it.first) {
            "forward" -> forward(it.second)
            "up" -> up(it.second)
            "down" -> down(it.second)
        }
    }
}
